docs/src/references/mappers/c8y-mapper.md

```diff
@@ -1325,3 +1325,191 @@ te/device/main///cmd/log_upload/<cmd_id>
+
+#### Phase 1
+
+**Live mapper**
+
+When the mapper is live, if telemetry messages for a new entity is received before its registration message,
+they are simply dropped until that registration message is processed, with the assumption that,
+telemetry data updates will repeat eventually and hence will make up for the lost ones.
+But, all non-telemetry data like twin updates, capability messages etc are cached in-memory,
+and will be processed and registered in the entity store once the registration message eventually arrives.
+Keeping them cached in-memory is not a big deal, as they are eventually going to end-up in the in-memory entity store anyway.
+
+**On mapper restart**
+
+Sticking to the same solution of dropping telemetry messages on a mapper restart might result in loosing too much data,
+when the volume of queued telemetry messages are very high on the broker,
+increasing the probability of a lot of those telemetry messages getting delivered to the mapper first,
+before their respective registration messages are.
+
+To avoid so much data loss, on restart the mapper will selectively subscribe to the registration topics first,
+process all the retained registration messages (with a timeout), after which the regular message processing will start.
+The timeout is required because MQTT offers no clear boundary/indication when all queued retained messages are delivered,
+and hence it has to be controlled with a deterministic waiting period.
+Since this timed waiting is during the startup, the impact won't be that visible to external users.
+
+**Auto registration**
+
+To avoid any conflicts between auto-registration and explicit registration,
+resulting in nested child devices from wrongly getting auto-registered as immediate child devices,
+we can just instruct the users to turn-off auto-registration if they have nested child devices in their deployment.
+Explicit registration must be used even for immediate child devices in such deployments.
+Or to say it differently, use auto-registration only in simple deployments where there are only immediate child devices.
+
+#### Phase 2
+
+To mitigate the telemetry message drop limitation of the above solution,
+cloud-specific MQTT topics can be used as secondary caches instead, for those "early messages".
+For e.g, when early measurement messages are received on `te/device/child1///m/environment`,
+they can be published back to `c8y-internal/device/child1///m/environment` topic and keep them queued there.
+Once the registration message is eventually processed, the mapper will dynamically subscribe to `c8y-internal/device/child1///+/+` topic
+and process those messages first, before the regular telemetry message on the `te/` topics are processed for it.
+The in-memory caching of non-telemetry data can also be replaced with this topic-based caching, but for minimal little gains.
+
+This solution requires some major changes in the current MQTT mapper to add support for dynamic subscriptions,
```

Comment:
We can also use an append-only log instead of MQTT topics to avoid the dynamic MQTT subscription, as reading from that log would be much easier than the MQTT topic.